### PR TITLE
Refactor flash attention implementation in transformers

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
EDIT: just refactor for now

 Enables us to run transformers model with Ragged Tensors: 

image

One of the goals is also to make it easy for people to re-define the `ExtraKwargs` typedict, to build on top of transformers

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @fxmarty, @LysandreJik and @OlivierDehaene

**Comment by raghukiran1224** [comment]
Is this using FA2 or SDPA?

**Comment by ArthurZucker** [comment]
FA2

File: src/transformers/flash_attention_utils.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        # TODO: Remove the `query_length != 1` check once Flash Attention for RoCm is bumped to 2.1. For details, please see the comment in LlamaFlashAttention2 __init__.
+        causal = is_causal and query_length != 1
+
+    use_sliding_windows = (
+        _flash_supports_window_size and sliding_window is not None and cache_position > sliding_window
+    )
+    flash_kwargs = {"window_size": (sliding_window, sliding_window)} if use_sliding_windows else {}
+
+    # Contains at least one padding token in the sequence
+    if attention_mask is not None:
```
Comment (fxmarty) [inline src/transformers/flash_attention_utils.py:115]: in TGI there's no `attention_mask`

Comment (fxmarty) [reply src/transformers/flash_attention_utils.py:115]: The ragged tensor should be put as an
```
elif all(["cu_seqlens", "max_seq_lens"]) in kwargs:
   # ...
```

File: src/transformers/flash_attention_utils.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            cu_seqlens_k=cu_seqlens_k,
+            max_seqlen_q=max_seqlen_in_batch_q,
+            max_seqlen_k=max_seqlen_in_batch_k,
+            dropout_p=dropout,
+            softmax_scale=softmax_scale,
+            causal=causal,
+            **flash_kwargs,
+        )
+        if unpad_inputs:
+            attn_output = _pad_input(attn_output_unpad, indices_q, batch_size, query_length)
```
Comment (fxmarty) [inline src/transformers/flash_attention_utils.py:144]: not in line with `_pad_input` signature

Comment (fxmarty) [inline src/transformers/flash_attention_utils.py:144]:
```suggestion
            attn_output = pad_input(attn_output_unpad, indices_q, batch_size, query_length)
```

File: src/transformers/models/llama/modeling_llama.py [hunk 3]
```diff
@@ -900,6 +793,7 @@ def forward(
         output_hidden_states: Optional[bool] = None,
         return_dict: Optional[bool] = None,
         cache_position: Optional[torch.LongTensor] = None,
+        **kwargs: Unpack[ExtraKwargs],
```
Comment (fxmarty) [inline src/transformers/models/llama/modeling_llama.py:796]: kwargs is missing for `LlamaForCausalLM.forward`

File: src/transformers/models/llama/modeling_llama.py [hunk 4]
```diff
@@ -464,8 +437,15 @@ def forward(
...
-            query_states, key_states, value_states, attention_mask, q_len, dropout=dropout_rate
+        attn_output = _flash_attention_forward(
+            query_states,
+            key_states,
+            value_states,
+            attention_mask,
+            q_len,
+            dropout=dropout_rate,
+            sliding_window=getattr(self, "sliding_window", None),
+            cache_position=cache_position,
```
Comment (fxmarty) [inline src/transformers/models/llama/modeling_llama.py:448]: `is_causal`, `_flash_attn_uses_top_left_mask` are not passed although they should

File: src/transformers/flash_attention_utils.py (hunk 1 above)
Comment (fxmarty) [reply src/transformers/flash_attention_utils.py:113]: outdated

File: src/transformers/models/gemma/modeling_gemma.py [hunk 5]
```diff
@@ -23,53 +23,34 @@
...
 
 import torch
-import torch.nn.functional as F
 import torch.utils.checkpoint
 from torch import nn
 from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
 
 from ...activations import ACT2FN
 from ...cache_utils import Cache, DynamicCache, StaticCache
+from ...flash_attention_utils import _flash_attention_forward
```
Comment (fxmarty) [inline src/transformers/models/gemma/modeling_gemma.py:32]: header says we should modify diff_gemma.py

File: src/transformers/flash_attention_utils.py
Comment (amyeroberts) [inline src/transformers/flash_attention_utils.py:1]: Should this sit under the utils folder?

File: src/transformers/flash_attention_utils.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
+import inspect
```
Comment (amyeroberts) [inline src/transformers/flash_attention_utils.py:1]: Missing copyright header

File: src/transformers/flash_attention_utils.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+from typing import Optional
+
+import torch
+import torch.nn.functional as F
+
+from .utils import (
+    is_flash_attn_2_available,
+)
```
Comment (amyeroberts) [inline src/transformers/flash_attention_utils.py:9]: nit

```suggestion
from .utils import is_flash_attn_2_available
```

File: src/transformers/flash_attention_utils.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    max_seqlen_in_batch = seqlens_in_batch.max().item()
+    cu_seqlens = F.pad(torch.cumsum(seqlens_in_batch, dim=0, dtype=torch.int32), (1, 0))
+    return (
+        indices,
+        cu_seqlens,
+        max_seqlen_in_batch,
+    )
+
+
+def _upad_input(query_layer, key_layer, value_layer, attention_mask, query_length):
```
Comment (amyeroberts) [inline src/transformers/flash_attention_utils.py:31]: These would benefit from docstrings, or at the very least some typing hints, as they're so ubiquitous

**Review by amyeroberts — APPROVED** [review]
❤️ 👏 kill 👏 that 👏 code 👏 ❤️

File: src/transformers/flash_attention_utils.py
Comment (fxmarty) [reply src/transformers/flash_attention_utils.py:1]: @amyeroberts I renamed as `modeling_flash_attention_utils.py` (as `modeling_attn_mask_utils.py` that was put at the top level). There are also `tokenization_utils.py`, `time_series_utils.py` etc at the top level.

If you'd prefer it to be moved to `utils/` or elsewhere let me know!

Comment (amyeroberts) [reply src/transformers/flash_attention_utils.py:1]: I think it's OK to leave as-is :)

**Review by ArthurZucker — COMMENTED** [review]
Thanks so much @fxmarty for going to the end of this!

**Comment by fxmarty** [comment]
No more flash attention tests fail here compared to main (on H100).
```
================================================================== short test summary info ==================================================================
FAILED tests/models/bark/test_modeling_bark.py::BarkSemanticModelTest::test_flash_attn_2_from_config - ValueError: Unrecognized configuration class <class 'transformers.models.bark.configuration_bark.BarkSemanticConfig'> for this kind of AutoModel: AutoMo...
FAILED tests/models/bark/test_modeling_bark.py::BarkCoarseModelTest::test_flash_attn_2_from_config - ValueError: Unrecognized configuration class <class 'transformers.models.bark.configuration_bark.BarkCoarseConfig'> for this kind of AutoModel: AutoMode...
FAILED tests/models/dpr/test_modeling_dpr.py::DPRModelTest::test_sdpa_can_dispatch_on_flash - RuntimeError: No available kernel. Aborting execution.
FAILED tests/models/gemma/test_modeling_gemma.py::GemmaIntegrationTest::test_model_2b_flash_attn - OSError: You are trying to access a gated repo.
FAILED tests/models/gemma2/test_modeling_gemma2.py::Gemma2ModelTest::test_flash_attn_2_equivalence - AssertionError: assert False
FAILED tests/models/gemma2/test_modeling_gemma2.py::Gemma2ModelTest::test_sdpa_can_dispatch_on_flash - RuntimeError: No available kernel. Aborting execution.
FAILED tests/models/gpt_neox/test_modeling_gpt_neox.py::GPTNeoXModelTest::test_flash_attn_2_generate_padding_right - AssertionError: False is not true
FAILED tests/models/idefics2/test_modeling_idefics2.py::Idefics2ModelTest::test_flash_attn_2_inference_equivalence_right_padding - ValueError: You are attempting to perform batched generation with padding_side='right' this may lead to unexpected behaviour for Flash Attention version...
FAILED tests/models/idefics2/test_modeling_idefics2.py::Idefics2ForConditionalGenerationModelTest::test_flash_attn_2_inference_equivalence_right_padding - ValueError: You are attempting to perform batched generation with padding_side='right' this may lead to unexpected behaviour for Flash Attention version...
FAILED tests/models/jamba/test_modeling_jamba.py::JambaModelTest::test_flash_attn_2_generate_padding_right - AssertionError: ValueError not raised
FAILED tests/models/jamba/test_modeling_jamba.py::JambaModelTest::test_sdpa_can_dispatch_on_flash - RuntimeError: No available kernel. Aborting execution.
FAILED tests/models/m2m_100/test_modeling_m2m_100.py::M2M100ModelTest::test_flash_attn_2_from_config - ValueError: Unrecognized configuration class <class 'transformers.models.m2m_100.configuration_m2m_100.M2M100Config'> for this kind of AutoModel: AutoMo...
FAILED tests/models/m2m_100/test_modeling_m2m_100.py::M2M100ModelIntegrationTests::test_flash_attn_2_seq_to_seq_generation - RuntimeError: FlashAttention only support fp16 and bf16 data type
FAILED tests/models/mixtral/test_modeling_mixtral.py::MixtralModelTest::test_flash_attn_2_generate_padding_right - AssertionError: ValueError not raised
FAILED tests/models/olmo/test_modeling_olmo.py::OlmoModelTest::test_flash_attn_2_generate_padding_right - AssertionError: False is not true
FAILED tests/models/phi3/test_modeling_phi3.py::Phi3ModelTest::test_flash_attn_2_generate_padding_right - AssertionError: False is not true
FAILED tests/models/qwen2/test_modeling_qwen2.py::Qwen2ModelTest::test_flash_attn_2_generate_padding_right - AssertionError: ValueError not raised
FAILED tests/models/qwen2/test_modeling_qwen2.py::Qwen2ModelTest::test_flash_attn_2_inference_equivalence - AssertionError: assert False
FAILED tests/models/qwen2_moe/test_modeling_qwen2_moe.py::Qwen2MoeModelTest::test_flash_attn_2_generate_padding_right - AssertionError: ValueError not raised
FAILED tests/models/qwen2_moe/test_modeling_qwen2_moe.py::Qwen2MoeModelTest::test_flash_attn_2_inference_equivalence - AssertionError: assert False
FAILED tests/models/stablelm/test_modeling_stablelm.py::StableLmModelTest::test_flash_attn_2_generate_padding_right - AssertionError: False is not true
FAILED tests/models/starcoder2/test_modeling_starcoder2.py::Starcoder2ModelTest::test_flash_attn_2_generate_padding_right - AssertionError: ValueError not raised
FAILED tests/models/unispeech/test_modeling_unispeech.py::UniSpeechRobustModelTest::test_flash_attn_2_inference_equivalence - RuntimeError: mat1 and mat2 must have the same dtype, but got Float and BFloat16
FAILED tests/models/unispeech/test_modeling_unispeech.py::UniSpeechRobustModelTest::test_flash_attn_2_inference_equivalence_right_padding - RuntimeError: mat1 and mat2 must have the same dtype, but got Float and BFloat16
FAILED tests/models/unispeech/test_modeling_unispeech.py::UniSpeechRobustModelTest::test_sdpa_can_dispatch_on_flash - RuntimeError: mat1 and mat2 must have the same dtype, but got Float and Half
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelTest::test_flash_attn_2_from_config - IndexError: too many indices for tensor of dimension 2
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelTest::test_flash_attn_2_inference_equivalence_right_padding - IndexError: too many indices for tensor of dimension 2
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperStandaloneDecoderModelTest::test_flash_attn_2_from_config - IndexError: too many indices for tensor of dimension 2
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperStandaloneDecoderModelTest::test_flash_attn_2_generate_left_padding - IndexError: too many indices for tensor of dimension 2
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperStandaloneDecoderModelTest::test_flash_attn_2_inference_equivalence - AssertionError: assert False
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperStandaloneDecoderModelTest::test_flash_attn_2_inference_equivalence_right_padding - AssertionError: assert False
================================= 31 failed, 1064 passed, 1789 skipped, 62671 deselected, 166 warnings in 331.97s (0:05:31) =================================
```

Testing on MI250 for extra safety and good to merge.

**Pull request merged**